packages/pg/lib/stream.js

```diff
@@ -0,0 +1,176 @@
+var { EventEmitter } = require('events')
+
+/**
+ * Get a socket stream compatible with the current runtime environment.
+ * @returns {Duplex}
+ */
+module.exports.getStream = function getStream(allowSsl) {
+  try {
+    var net = require('net')
+    if (typeof net.Socket === 'function') {
+      return new net.Socket()
+    }
+  } catch (e) {
+    // `net` module is not available
+  }
+  return new CFSocket(allowSsl)
+}
+
+module.exports.getSecureStream = function getSecureStream(options) {
+  var tls = require('tls')
+  if (tls.connect) {
+    return tls.connect(options)
+  } else {
+    options.socket.startTls(options)
+    return options.socket
+  }
+}
+
+/**
+ * Wrapper around the Cloudflare built-in socket that can be used by the `Connection`.
+ */
+class CFSocket extends EventEmitter {
+  constructor(ssl) {
+    super()
+
+    this.ssl = ssl
+
+    this.writable = false
+    this.destroyed = false
+
+    this._upgrading = false
+    this._cfSocket = null
+    this._cfWriter = null
+    this._cfReader = null
+  }
+
+  setNoDelay() {
+    return this
+  }
+  setKeepAlive() {
+    return this
+  }
+  ref() {
+    return this
+  }
+  unref() {
+    return this
+  }
+
+  async connect(port, host, connectListener) {
+    try {
+      log('connecting')
+      if (connectListener) this.once('connect', connectListener)
+
+      const options = this.ssl ? { secureTransport: 'allow' } : {}
+      const { connect } = await import('cloudflare:sockets')
+      this._cfSocket = connect(`${host}:${port}`, options)
+      this._cfWriter = this._cfSocket.writable.getWriter()
+
+      this._cfSocket.closed.then(() => {
+        if (!this._upgrading) {
+          log('CF socket closed')
+          this._cfSocket = null
+          this.emit('close')
```

Comment:
If any of the registered `close` event handlers throw an error synchronously, the promise returned by this then call is going to register an unhandled rejection. It would be good to attach a catch handler to this that forwards the error to an `error` event on this class.